src/annotator/highlight-clusters.tsx

```diff
@@ -121,6 +141,105 @@ export class HighlightClusterController implements Destroyable {
     this._activate(this._isActive());
   }
 
+  /**
+   * Update highlight elements — <hypothesis-highlight> elements and SVG <rect>s
+   * with data about their nesting depth, and order SVG <rect>s such that
+   * inner (nested) highlights are stacked on top of outer highlights.
+   */
+  _updateClusters() {
+    if (!this._isActive()) {
+      /* istanbul ignore next */
+      return;
+    }
+
+    this._updateHighlightData(getHighlights(this._element));
+    this._updateSVGHighlightOrdering();
+  }
+
+  /**
+   * Walk a tree of <hypothesis-highlight> elements, adding `data-nesting-level`
+   * and `data-cluster-level` data attributes to <hypothesis-highlight>s and
+   * their associated SVG highlight (<rect>) elements.
+   *
+   * - `data-nesting-level` - generational depth of the applicable
+   *   `<hypothesis-highlight>` relative to outermost `<hypothesis-highlight>`.
+   * - `data-cluster-level` - number of `<hypothesis-highlight>` generations
+   *   since the cluster value changed.
+   *
+   * @param highlightEls - A collection of sibling <hypothesis-highlight>
+   * elements
+   * @param parentCluster - The cluster value of the parent highlight to
+   * `highlightEls`, if any
+   * @param nestingLevel - The nesting "level", relative to the outermost
+   * <hypothesis-highlight> element (0-based)
+   * @param parentClusterLevel - The parent's nesting depth, per its cluster
+   * value (`parentCluster`). i.e. How many levels since the cluster value
+   * changed? This allows for nested styling of highlights of the same cluster
+   * value.
+   */
+  _updateHighlightData(
+    highlightEls: HighlightElement[],
+    parentCluster = '',
+    nestingLevel = 0,
+    parentClusterLevel = 0
+  ) {
+    Array.from(highlightEls).forEach(hEl => {
```

Comment:
I would generally suggest to use a `for..of` loop rather than `forEach` except at the end of a chain of filter/map etc, as it allows use of `return`/`continue`/`break` more easily if that is useful. Also the loop variable can be marked `const` if you're not going to change it.